Async support for trace decorator

MPKonst opened the pull request:
### Related Issues/PRs
Resolves issue #12756 .
<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve -->

### What changes are proposed in this pull request?

This PR extends the `@trace` decorator to support decorating async functions. 

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

I think this change does not require an update to the documentation but I'd be happy to add something, if you think otherwise.

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

The `@trace` decorator now supports async functions.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

MPKonst committed:
Add async support to the @trace decorator

    - @trace decorator now checks whether the function its wrapping is async and, if so, returns an async wrapper
    - add a couple of async tests for the new functionality
    - also add pytest-asyncio to the requirements, so we can run async tests

Signed-off-by: Momchil Konstantinov <konstantinov.momchil@gmail.com>

MPKonst committed:
Merge branch 'master' into async_support_for_trace_decorator

Signed-off-by: Momchil Konstantinov <konstantinov.momchil@gmail.com>

MPKonst committed:
Remove commented line

Signed-off-by: Momchil Konstantinov <konstantinov.momchil@gmail.com>

MPKonst committed:
Remove superfluous line.

Signed-off-by: Momchil Konstantinov <konstantinov.momchil@gmail.com>

B-Step62 commented on tests/tracing/test_fluent.py:
```diff
@@ -119,6 +137,67 @@ def test_trace(with_active_run):
     }
 
 
+@pytest.mark.parametrize("with_active_run", [True, False])
```
nit: I'd just add a new parameter in `test_trace` to ensure the assertions are exactly same between sync and async tests.

MPKonst replied:
Good point. That was my original thought but then I wanted to make the test itself async. Your comment made me realise that I can just wrap the call in an `asyncio.run`, so that's what I've done now.

B-Step62 commented on requirements/test-requirements.txt:
```diff
@@ -6,6 +6,7 @@ pytest==8.1.0
 pytest-cov
 pytest-timeout
 pytest-localserver==0.5.0
+pytest-asyncio==0.23.8
```
Do we need to pin asyncio to this version?

MPKonst replied:
Removed this line, as we pytest-asyncio not needed anymore.

B-Step62 reviewed (commented):
Thanks for the PR @MPKonst! The approach looks beautiful to me and I tested it works with simple cases🙂
Can you mark the PR "ready for review" so those CI tests are unblocked? Once all tests pass, I think we are good to go!

MPKonst committed:
Add sync/async parametrization to tests.

  - remove pytest-asyncio from requirements, as we use asyncio.run in the tests now.

Signed-off-by: Momchil Konstantinov <konstantinov.momchil@gmail.com>

MPKonst committed:
Merge branch 'master' into async_support_for_trace_decorator

Signed-off-by: Momchil Konstantinov <konstantinov.momchil@gmail.com>

MPKonst committed:
Remove asyncio config from pytest.ini

Signed-off-by: Momchil Konstantinov <konstantinov.momchil@gmail.com>

B-Step62 reviewed (approved):
The implementation looks solid and it works beautifully! Thank you so much for the great contribution @MPKonst!!😄
